akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala

```diff
@@ -202,30 +204,29 @@ private[akka] abstract class InternalProjectionState[Offset, Envelope](
           () => handler.process(envelopes))
       }
 
-      // FIXME create a SourceProvider trait that implies mergeable offsets?
-      if (sourceProvider.isOffsetMergeable) {
-        val batches = envelopesAndOffsets.groupBy {
-          // FIXME matched source provider should always be MergeableOffset
-          case ProjectionContextImpl(offset: MergeableOffset[_, _], _) =>
-            // FIXME we can assume there's only one actual offset per envelope, but there should be a better way to represent this
-            val mergeableKey = offset.entries.head._1.asInstanceOf[MergeableKey]
-            mergeableKey.surrogateKey
-          case _ =>
-            // should never happen
-            throw new IllegalStateException("The offset should always be of type MergeableOffset")
-        }
-
-        // process batches in sequence, but not concurrently, in order to provide singled threaded guarantees
-        // to the user envelope handler
-        serialize(batches) { (surrogateKey, partitionedEnvelopes) =>
-          logger.debug("Processing grouped envelopes for MergeableOffset with key [{}]", surrogateKey)
-          processEnvelopes(partitionedEnvelopes)
-        }
-
-      } else {
-        processEnvelopes(envelopesAndOffsets)
+      sourceProvider match {
+        case _: MergeableOffsetSourceProvider[_, Offset, Envelope] =>
+          val batches = envelopesAndOffsets
+            .asInstanceOf[immutable.Seq[ProjectionContextImpl[MergeableOffset[MergeableKey, _], Envelope]]]
+            .flatMap { context => context.offset.entries.toSeq.map { case (key, _) => (key, context) } }
+            .groupBy { case (key, _) => key }
+            .map {
+              case (key, keyAndContexts) =>
+                val envs = keyAndContexts.map { case (_, context) => context }
+                key.surrogateKey -> envs
+            }
+            .asInstanceOf[Map[String, immutable.Seq[ProjectionContextImpl[Offset, Envelope]]]]
```

Comment:
can we have this cast somewhere in the inner `map` instead?

Reply:
:+1: